apollo-federation/src/query_graph/graph_path.rs

```diff
@@ -3539,6 +3583,91 @@ impl TryFrom<&'_ OpPath> for Vec<QueryPathElement> {
     }
 }
 
+pub(crate) fn concat_paths_in_parents(
+    first: &Option<Arc<OpPath>>,
+    second: &Option<Arc<OpPath>>,
+) -> Option<Arc<OpPath>> {
+    if let (Some(first), Some(second)) = (first, second) {
+        Some(Arc::new(concat_op_paths(first.deref(), second.deref())))
+    } else {
+        None
+    }
+}
+
+pub(crate) fn concat_op_paths(head: &OpPath, tail: &OpPath) -> OpPath {
+    // While this is mainly a simple array concatenation, we optimize slightly by recognizing if the
+    // tail path starts by a fragment selection that is useless given the end of the head path
+    if head.is_empty() {
+        return tail.clone();
+    }
+    let mut result = head.clone();
+    if tail.is_empty() {
+        return result;
+    }
+    let conditionals = head.conditional_directives();
+    let last_of_head = head.last().unwrap();
+    let mut tail_index = 0;
+    let tail_path = tail.0.clone();
+
+    // Note that in practice, we may be able to eliminate a few elements at the beginning of the path
+    // due do conditionals ('@skip' and '@include'). Indeed, a (tail) path crossing multiple conditions
+    // may start with: [ ... on X @include(if: $c1), ... on X @skip(if: $c2), (...)], but if `head`
+    // already ends on type `X` _and_ both the conditions on `$c1` and `$c2` are already found on `head`,
+    // then we can remove both fragments in `tail`.
+    while tail_index < tail_path.len()
+        && is_useless_followup_element(last_of_head, &tail_path[tail_index], &conditionals)
+            .is_ok_and(|is_useless| is_useless)
+    {
+        tail_index += 1;
+    }
+    if tail_index < tail_path.len() {
+        while tail_index < tail_path.len() {
+            result.0.push(tail_path[tail_index].clone());
+            tail_index += 1;
+        }
+    }
+    result
+}
+
+fn is_useless_followup_element(
+    first: &OpPathElement,
+    followup: &OpPathElement,
+    conditionals: &DirectiveList,
+) -> Result<bool, FederationError> {
+    let type_of_first: Option<CompositeTypeDefinitionPosition> = match first {
+        OpPathElement::Field(field) => Some(field.data().output_base_type()?.try_into()?),
+        OpPathElement::InlineFragment(fragment) => fragment.data().type_condition_position.clone(),
+    };
+
+    let Some(type_of_first) = type_of_first else {
+        return Ok(false);
+    };
+
+    // The followup is useless if it's a fragment (with no directives we would want to preserve) whose type
+    // is already that of the first element (or a supertype).
+    return match followup {
+        OpPathElement::Field(_) => Ok(false),
```

Comment:
From the JS code follow up element is useless only if its a fragment with no directives and same type condition as the type of the first element (or a supertype)

```js
  return !!typeOfFirst     // already checked above 
    && followup.kind === 'FragmentElement' // <==== this can only be true for fragments
    && !!followup.typeCondition
    && (followup.appliedDirectives.length === 0 || isDirectiveApplicationsSubset(conditionals, followup.appliedDirectives))
    && isSubtype(followup.typeCondition, typeOfFirst);
```

Reply:
gotcha gotcha! that match nested match return is ... complicated... to read through